File: scripts/test-connect-api-opening-templates.rb
# frozen_string_literal: true

require 'minitest/autorun'
require_relative '../app/services/connect_api/opening_template_catalog'

class OpeningTemplateCatalogTest < Minitest::Test
  Catalog = ConnectApi::OpeningTemplateCatalog

  def remote(name = 'hello', language = 'pt_BR', **attributes)
    {
      'id' => "remote-#{name}-#{language}", 'name' => name, 'language' => language,
      'status' => 'APPROVED', 'category' => 'UTILITY',
      'components' => [{ 'type' => 'BODY', 'text' => "Remote content for #{name}" }]
    }.merge(attributes.transform_keys(&:to_s))
  end

  def catalog(templates = [], instance = 'instance-a')
    Catalog.new(templates, instance_name: instance)
  end

  def test_imports_remote_content_verbatim_and_enables_only_exact_hello
    input = [remote, remote('Hello'), remote('hello_world'), remote('notice')]
    result = catalog.reconcile(input)
    assert_equal [true, false, false, false], result.map { |t| t['hub_opening_enabled'] }
    assert_equal input.map { |t| t['components'] }, result.map { |t| t['components'] }
    refute input.first.key?('hub_opening_enabled')
  end

  def test_never_synthesizes_hello_or_other_templates
    assert_empty catalog.reconcile([])
    assert_equal ['notice'], catalog.reconcile([remote('notice')]).map { |t| t['name'] }
  end

  def test_preserves_disable_hello_and_enable_other_choices
    entries = catalog.reconcile([remote, remote('notice')])
    entries = catalog(entries).set_enabled(name: 'hello', language: 'pt_BR', enabled: false)
    entries = catalog(entries).set_enabled(name: 'notice', language: 'pt_BR', enabled: true)
    result = catalog(entries).reconcile([remote, remote('notice'), remote('new')])
    assert_equal [false, true, false], result.map { |t| t['hub_opening_enabled'] }
  end

  def test_disappeared_templates_keep_choices_but_cannot_be_used
    entries = catalog.reconcile([remote])
    result = catalog(entries).reconcile([])
    assert result.first['hub_opening_enabled']
    refute result.first['hub_remote_present']
    assert_empty catalog(result).available_templates(opening_only: true)
    result = catalog(result).reconcile([remote])
    assert_equal ['hello'], catalog(result).available_templates(opening_only: true).map { |t| t['name'] }
  end

  def test_disabled_choice_survives_disappearance_and_remote_id_change
    entries = catalog.reconcile([remote])
    entries = catalog(entries).set_enabled(name: 'hello', language: 'pt_BR', enabled: false)
    missing = catalog(entries).reconcile([])
    result = catalog(missing).reconcile([remote(id: 'new-remote-id')])
    assert_equal 'new-remote-id', result.first['id']
    refute result.first['hub_opening_enabled']
  end

  def test_rejected_pending_paused_and_disabled_are_unavailable
    %w[REJECTED PENDING PAUSED DISABLED DELETED UNKNOWN].each do |status|
      result = catalog.reconcile([remote(status: status)])
      assert result.first['hub_opening_enabled']
      assert_empty catalog(result).available_templates, status
    end
  end

  def test_approved_status_is_case_insensitive_and_missing_status_is_not_fabricated
    assert catalog.reconcile([remote(status: 'approved')]).first['hub_remote_available']
    template = remote.reject { |key, _| key == 'status' }
    result = catalog.reconcile([template])
    assert result.first['hub_remote_available']
    refute result.first.key?('status')
  end

  def test_remote_unavailability_overrides_the_local_choice
    [false, 0, 'false', '0'].each do |value|
      %w[available enabled].each do |key|
        result = catalog.reconcile([remote(**{ key => value })])
        assert_empty catalog(result).available_templates
      end
    end
  end

  def test_languages_are_independent
    entries = catalog.reconcile([remote, remote('hello', 'en_US')])
    entries = catalog(entries).set_enabled(name: 'hello', language: 'en_US', enabled: false)
    assert_equal ['pt_BR'], catalog(entries).available_templates(opening_only: true).map { |t| t['language'] }
  end

  def test_instances_and_inboxes_do_not_share_choices
    first = catalog.reconcile([remote, remote('notice')])
    first = catalog(first).set_enabled(name: 'notice', language: 'pt_BR', enabled: true)
    assert_empty catalog(first, 'instance-b').available_templates(opening_only: true)
    second = catalog(first, 'instance-b').reconcile([remote('notice')])
    refute catalog(second, 'instance-b').entries.first['hub_opening_enabled']
    assert catalog(second).available_templates(opening_only: true).any? { |t| t['name'] == 'notice' }
  end

  def test_legacy_or_unverified_templates_are_not_offered_or_retained_as_remote_records
    assert_empty catalog([remote]).available_templates
    assert_empty catalog([remote]).reconcile([])
    assert_raises(Catalog::TemplateNotFound) do
      catalog([remote]).set_enabled(name: 'hello', language: 'pt_BR', enabled: true)
    end
  end

  def test_remote_payload_cannot_override_administrative_flags
    result = catalog.reconcile([remote('notice', hub_opening_enabled: true, hub_instance_name: 'other')])
    refute result.first['hub_opening_enabled']
    assert_equal 'instance-a', result.first['hub_instance_name']
  end

  def test_invalid_response_is_rejected_instead_of_being_imported_as_empty
    [nil, {}, 'oops', [nil], [{ 'name' => 'hello' }], [remote(components: nil)]].each do |input|
      assert_raises(Catalog::InvalidTemplate) { catalog.reconcile(input) }
    end
  end

  def test_invalid_toggle_and_unknown_template_are_rejected
    entries = catalog.reconcile([remote])
    [nil, '', 'false', 'true', 0, 1].each do |enabled|
      assert_raises(ArgumentError) { catalog(entries).set_enabled(name: 'hello', language: 'pt_BR', enabled: enabled) }
    end
    assert_raises(Catalog::TemplateNotFound) do
      catalog(entries).set_enabled(name: 'other', language: 'pt_BR', enabled: true)
    end
  end

  def test_reply_catalog_can_include_a_disabled_opening_template
    entries = catalog.reconcile([remote('notice')])
    assert_equal 1, catalog(entries).available_templates.size
    assert_empty catalog(entries).available_templates(opening_only: true)
  end

  def test_a_status_change_updates_availability_without_overwriting_the_preference
    entries = catalog.reconcile([remote])
    paused = catalog(entries).reconcile([remote(status: 'PAUSED')])
    assert_empty catalog(paused).available_templates(opening_only: true)
    approved = catalog(paused).reconcile([remote])
    assert_equal 1, catalog(approved).available_templates(opening_only: true).size
  end
end

File: scripts/test-connect-api-opening-templates.mjs
// Dependency-free tests against the actual Vue options/getters, not a reimplementation.
// These do not replace a production bundle or browser visual validation.
import assert from 'node:assert/strict';
import { test } from 'node:test';
import { readFileSync } from 'node:fs';
import vm from 'node:vm';

const root = new URL('../', import.meta.url);
const read = path => readFileSync(new URL(path, root), 'utf8');
const base = 'app/javascript/dashboard/';
const settingsPath = `${base}routes/dashboard/settings/inbox/settingsPage/`;
const contactPath = `${base}routes/dashboard/conversation/contact/`;
const callPath = `${base}components/widgets/conversation/ConnectApiCallPanelPolished.vue`;
const script = text => text.match(/<script[^>]*>([\s\S]*?)<\/script>/)?.[1] ?? text;
const noImports = text => text.replace(/^import[\s\S]*?from ['"][^'"]+['"];\s*/gm, '');
function component(path, globals = {}) {
  const code = noImports(script(read(path))).replace('export default', 'globalThis.component =');
  const context = { console, mapGetters: () => ({}), ...globals };
  vm.runInNewContext(code, context, { filename: path });
  return context.component;
}
const moduleText = read(`${base}store/modules/inboxes.js`);
const getterCode = moduleText.slice(moduleText.indexOf('export const getters ='), moduleText.indexOf('export const actions ='));
const context = { INBOX_TYPES: {}, Array };
vm.runInNewContext(getterCode.replace('export const getters =', 'globalThis.getters ='), context);
const getters = context.getters;
const template = (name = 'hello', language = 'pt_BR') => ({ name, language, status: 'APPROVED', components: [{ type: 'BODY', text: 'From API' }] });
const names = values => Array.from(values, item => item.name);

test('opening uses the server-filtered per-inbox catalog, not all templates', () => {
  const state = { records: [{ id: 1, provider: 'connectapi', message_templates: [template(), template('aviso')], opening_templates: [template('aviso')] }] };
  assert.deepEqual(names(getters.getWhatsAppTemplates(state)(1, true)), ['aviso']);
  assert.deepEqual(names(getters.getWhatsAppTemplates(state)(1)), ['hello', 'aviso']);
});
test('missing/empty opening catalog fails closed without fallback to additional attributes', () => {
  for (const opening_templates of [undefined, null, []]) {
    const state = { records: [{ id: 1, provider: 'connectapi', opening_templates, message_templates: [template()], additional_attributes: { message_templates: [template()] } }] };
    assert.equal(getters.getWhatsAppTemplates(state)(1, true).length, 0);
  }
});
test('one inbox cannot select another inbox catalog', () => {
  const state = { records: [{ id: 1, provider: 'connectapi', opening_templates: [template('one')] }, { id: 2, provider: 'connectapi', opening_templates: [template('two')] }] };
  assert.deepEqual(names(getters.getWhatsAppTemplates(state)(2, true)), ['two']);
  assert.equal(getters.getWhatsAppTemplates(state)(3, true).length, 0);
});
test('Meta templates keep their original catalog path', () => {
  const state = { records: [{ id: 1, provider: 'whatsapp_cloud', message_templates: [template('meta')], opening_templates: [] }] };
  assert.deepEqual(names(getters.getWhatsAppTemplates(state)(1, true)), ['meta']);
});
test('unsupported components do not crash or appear in the picker', () => {
  const state = { records: [{ id: 1, provider: 'connectapi', opening_templates: [template(), { name: 'bad', components: null }, { ...template('image'), components: [{ format: 'IMAGE' }] }] }] };
  assert.deepEqual(names(getters.getWhatsAppTemplates(state)(1, true)), ['hello']);
});
test('picker filters pending/rejected templates and preserves absent statuses only for Connect API', () => {
  const picker = component(`${base}components/widgets/conversation/WhatsappTemplates/TemplatesPicker.vue`);
  const all = [template(), { ...template('pending'), status: 'PENDING' }, { ...template('nostatus'), status: undefined }, { ...template('nobody'), components: [] }];
  const self = { inboxId: 1, openingOnly: true, isConnectApi: true, $store: { getters: { 'inboxes/getWhatsAppTemplates': (id, only) => {
    assert.equal(id, 1); assert.equal(only, true); return all;
  } } } };
  assert.deepEqual(names(picker.computed.whatsAppTemplateMessages.call(self)), ['hello', 'nostatus']);
  self.isConnectApi = false;
  assert.deepEqual(names(picker.computed.whatsAppTemplateMessages.call(self)), ['hello']);
});
test('changing inbox clears a selected opening template', () => {
  const wrapper = component(`${contactPath}WhatsappTemplates.vue`, { TemplatesPicker: {}, TemplateParser: {} });
  const events = [];
  const self = { selectedWaTemplate: template(), $emit: (...args) => events.push(args) };
  self.onResetTemplate = wrapper.methods.onResetTemplate.bind(self);
  wrapper.watch.inboxId.call(self);
  assert.equal(self.selectedWaTemplate, null);
  assert.equal(events[0][1], false);
  assert.match(read(`${contactPath}WhatsappTemplates.vue`), /opening-only/);
});
test('new conversation cannot fall back to free text just because Connect API returns no templates', () => {
  const code = script(read(`${contactPath}ConversationForm.vue`));
  const body = code.match(/hasWhatsappTemplates\(\) \{([\s\S]*?)\n    \}/)[1];
  const hasTemplates = new Function(body);
  assert.equal(hasTemplates.call({ selectedInbox: { inbox: { provider: 'connectapi', message_templates: null } } }), true);
  assert.equal(hasTemplates.call({ selectedInbox: { inbox: { provider: 'email' } } }), false);
});
test('admin load ignores late results from a different inbox', async () => {
  let resolve;
  const admin = component(`${settingsPath}ConnectApiOpeningTemplates.vue`, { InboxAPI: { getOpeningTemplates: () => new Promise(done => { resolve = done; }) }, useAlert: () => {} });
  const self = { ...admin.data(), inbox: { id: 1 } };
  self.applyCatalog = admin.methods.applyCatalog.bind(self);
  const pending = admin.methods.loadTemplates.call(self);
  self.inbox = { id: 2 };
  resolve({ data: { payload: [template('wrong-inbox')] } });
  await pending;
  assert.equal(self.templates.length, 0);
});
test('failed save restores the switch and does not mutate the persisted preference', async () => {
  const admin = component(`${settingsPath}ConnectApiOpeningTemplates.vue`, { InboxAPI: {}, useAlert: () => {} });
  const persisted = { ...template(), hub_opening_enabled: false };
  const self = { ...admin.data(), inbox: { id: 1 }, templates: [persisted], $store: { dispatch: async () => { throw new Error('network'); } } };
  self.key = admin.methods.key;
  const event = { target: { checked: true } };
  await admin.methods.setEnabled.call(self, persisted, event);
  assert.equal(event.target.checked, false);
  assert.equal(persisted.hub_opening_enabled, false);
  assert.equal(self.savingKey, null);
});
test('disconnect sends the existing payload once and always resets its busy state', async () => {
  const settings = component(`${settingsPath}ConnectApiConfiguration.vue`, { ConnectApiOpeningTemplates: {}, useAlert: () => {} });
  let resolve;
  let calls = 0;
  const self = { isDisconnecting: false, isReconciling: false, update: payload => {
    assert.equal(payload.disconnect, true); assert.equal(payload.connect, false); calls += 1;
    return new Promise(done => { resolve = done; });
  } };
  const first = settings.methods.disconnect.call(self);
  await settings.methods.disconnect.call(self);
  assert.equal(calls, 1);
  assert.equal(self.isDisconnecting, true);
  resolve(false);
  await first;
  assert.equal(self.isDisconnecting, false);
});
test('reconcile refreshes the real admin catalog only after a successful backend update', async () => {
  const settings = component(`${settingsPath}ConnectApiConfiguration.vue`, { ConnectApiOpeningTemplates: {}, useAlert: () => {} });
  let loads = 0;
  const self = { isReconciling: false, isDisconnecting: false, update: async () => true, $refs: { openingTemplates: { loadTemplates: async () => { loads += 1; } } } };
  await settings.methods.reconcile.call(self);
  assert.equal(loads, 1);
  self.update = async () => false;
  await settings.methods.reconcile.call(self);
  assert.equal(loads, 1);
  assert.equal(self.isReconciling, false);
});
test('mute and disconnect retain their actions with contained light/dark visual states', () => {
  const calls = read(callPath);
  const button = calls.match(/<button\s+v-if="mediaCallId === callId\(primaryCall\)"[\s\S]*?<\/button>/)?.[0];
  assert.ok(button);
  for (const token of ['rounded-xl', 'px-4', 'py-2.5', 'dark:bg-slate-700', ':aria-pressed', "action(primaryCall, 'mute')", 'animate-spin']) assert.ok(button.includes(token), token);
  const settings = read(`${settingsPath}ConnectApiConfiguration.vue`);
  const disconnect = settings.match(/<button[^>]*[\s\S]*?@click="disconnect"[\s\S]*?<\/button>/)?.[0];
  assert.ok(disconnect?.includes('bg-red-600'));
  assert.ok(disconnect?.includes('animate-spin'));
});
test('modified Vue/JavaScript scripts have valid JavaScript syntax', () => {
  const paths = [
    `${base}api/inboxes.js`, `${base}store/modules/inboxes.js`, callPath,
    `${base}components/widgets/conversation/WhatsappTemplates/TemplatesPicker.vue`,
    `${settingsPath}ConnectApiConfiguration.vue`, `${settingsPath}ConnectApiOpeningTemplates.vue`,
    `${contactPath}ConversationForm.vue`, `${contactPath}WhatsappTemplates.vue`,
  ];
  for (const path of paths) {
    const code = noImports(script(read(path))).replace(/export default/g, 'globalThis.result =').replace(/export const/g, 'const');
    assert.doesNotThrow(() => new vm.Script(code, { filename: path }), path);
  }
});
